File: templates/minimal-harness.ts
/**
 * Minimal Harness — Hera Architecture Reference
 *
 * Agent Harness is the orchestration layer. It wraps the agent loop with:
 * - Session management
 * - Compaction
 * - Skills & prompt templates
 * - Hook system
 * - Queue management (steer, follow-up, next-turn)
 *
 * Based on packages/agent/src/harness/agent-harness.ts
 */

// ============================================================================
// Types (simplified)
// ============================================================================

interface Message {
  role: "user" | "assistant" | "toolResult";
  content: unknown[];
  timestamp: number;
}

interface Tool {
  name: string;
  description: string;
  execute: (args: Record<string, unknown>) => Promise<string>;
}

interface Model {
  id: string;
  provider: string;
  contextWindow: number;
}

type HarnessPhase = "idle" | "turn" | "compact" | "fork" | "switch";

// ============================================================================
// Queue
// ============================================================================

class MessageQueue {
  private messages: Message[] = [];
  private mode: "all" | "one-at-a-time";

  constructor(mode: "all" | "one-at-a-time" = "one-at-a-time") {
    this.mode = mode;
  }

  enqueue(message: Message): void {
    this.messages.push(message);
  }

  drain(): Message[] {
    if (this.mode === "all") {
      const drained = [...this.messages];
      this.messages = [];
      return drained;
    }
    if (this.messages.length === 0) return [];
    return [this.messages.shift()!];
  }

  hasItems(): boolean {
    return this.messages.length > 0;
  }

  clear(): void {
    this.messages = [];
  }
}

// ============================================================================
// Session (simplified)
// ============================================================================

class SimpleSession {
  private messages: Message[] = [];

  appendMessage(message: Message): void {
    this.messages.push(message);
  }

  getMessages(): Message[] {
    return [...this.messages];
  }

  clear(): void {
    this.messages = [];
  }
}

// ============================================================================
// Agent Harness
// ============================================================================

interface HarnessOptions {
  model: Model;
  systemPrompt: string | ((context: { model: Model }) => string);
  tools: Tool[];
  session?: SimpleSession;
}

class AgentHarness {
  private model: Model;
  private systemPrompt: string | ((context: { model: Model }) => string);
  private tools: Map<string, Tool>;
  private session: SimpleSession;
  private phase: HarnessPhase = "idle";
  private steerQueue: MessageQueue;
  private followUpQueue: MessageQueue;
  private nextTurnQueue: Message[] = [];

  constructor(options: HarnessOptions) {
    this.model = options.model;
    this.systemPrompt = options.systemPrompt;
    this.tools = new Map(options.tools.map((t) => [t.name, t]));
    this.session = options.session ?? new SimpleSession();
    this.steerQueue = new MessageQueue("one-at-a-time");
    this.followUpQueue = new MessageQueue("one-at-a-time");
  }

  // --- Core API ---

  async prompt(text: string): Promise<string> {
    if (this.phase !== "idle") throw new Error("Harness is busy");
    this.phase = "turn";

    try {
      // Create user message
      const userMessage: Message = {
        role: "user",
        content: [{ type: "text", text }],
        timestamp: Date.now(),
      };

      // Prepend next-turn queue
      const messages = [...this.nextTurnQueue.splice(0), userMessage];

      // Build context
      const context = this.buildContext(messages);

      // Run agent loop (simplified)
      const response = await this.runAgentLoop(context);

      // Save to session
      this.session.appendMessage(userMessage);
      this.session.appendMessage(response);

      return this.extractText(response);
    } finally {
      this.phase = "idle";
    }
  }

  // --- Queue API ---

  steer(text: string): void {
    if (this.phase === "idle") throw new Error("Cannot steer while idle");
    this.steerQueue.enqueue({
      role: "user",
      content: [{ type: "text", text }],
      timestamp: Date.now(),
    });
  }

  followUp(text: string): void {
    this.followUpQueue.enqueue({
      role: "user",
      content: [{ type: "text", text }],
      timestamp: Date.now(),
    });
  }

  // --- Session API ---

  getSession(): SimpleSession {
    return this.session;
  }

  // --- State ---

  getPhase(): HarnessPhase {
    return this.phase;
  }

  // --- Internal ---

  private buildContext(additionalMessages: Message[]): { systemPrompt: string; messages: Message[]; tools: Tool[] } {
    const resolvedPrompt =
      typeof this.systemPrompt === "string" ? this.systemPrompt : this.systemPrompt({ model: this.model });

    return {
      systemPrompt: resolvedPrompt,
      messages: [...this.session.getMessages(), ...additionalMessages],
      tools: [...this.tools.values()],
    };
  }

  private async runAgentLoop(context: {
    systemPrompt: string;
    messages: Message[];
    tools: Tool[];
  }): Promise<Message> {
    // Simplified agent loop — in production, this calls the LLM
    const lastMessage = context.messages[context.messages.length - 1];
    const text = (lastMessage?.content?.[0] as any)?.text ?? "";

    // Simulate LLM response
    return {
      role: "assistant",
      content: [{ type: "text", text: `Response to: "${text}"` }],
      timestamp: Date.now(),
    };
  }

  private extractText(message: Message): string {
    return message.content
      .filter((c: any) => c.type === "text")
      .map((c: any) => c.text)
      .join("");
  }
}

// ============================================================================
// Example Usage
// ============================================================================

async function main() {
  const harness = new AgentHarness({
    model: { id: "gpt-4", provider: "openai", contextWindow: 128000 },
    systemPrompt: "You are a helpful coding assistant.",
    tools: [
      {
        name: "read",
        description: "Read a file",
        execute: async (args) => `Contents of ${args.path}`,
      },
    ],
  });

  // Basic prompt
  const response1 = await harness.prompt("Hello, what can you do?");
  console.log("Response 1:", response1);

  // Follow-up (queued, runs after agent would stop)
  harness.followUp("Now explain the architecture");

  // Steer (inject mid-run — only works while agent is running)
  // harness.steer("Focus on the session system");

  // Check state
  console.log("Phase:", harness.getPhase());
  console.log("Session messages:", harness.getSession().getMessages().length);
}

main().catch(console.error);

File: templates/python/minimal_harness.py
"""
Minimal Harness — Hera Architecture Reference (Python)

Agent harness: the top-level orchestrator that wires everything together.
Manages agent lifecycle, tool registration, extension loading, and session handling.

Based on Pi Agent's agent-harness.ts (packages/agent/src/agent-harness.ts).
"""

from __future__ import annotations

import asyncio
import signal
import sys
from dataclasses import dataclass, field
from typing import Any

from minimal_agent_loop import AgentContext, Message, TextContent, agent_loop
from minimal_extension import Extension, ExtensionManager, HookPoint
from minimal_provider import LLMProvider, OpenAIProvider, ProviderConfig
from minimal_session import InMemoryStorage, Session
from minimal_tool import BaseTool, ToolRegistry


# ============================================================================
# Config
# ============================================================================

@dataclass
class AgentConfig:
    """Configuration for the agent harness."""
    name: str = "my-agent"
    system_prompt: str = "You are a helpful coding assistant."
    max_context_tokens: int = 128000
    max_tool_retries: int = 3
    log_level: str = "info"


# ============================================================================
# Agent Harness
# ============================================================================

class AgentHarness:
    """
    Top-level orchestrator. Wires together:
    - LLM Provider (OpenAI/Anthropic/local)
    - Tools (read, write, bash, etc.)
    - Extensions (logging, security, timing)
    - Session (history, branching, persistence)
    """

    def __init__(self, config: AgentConfig, provider: LLMProvider):
        self.config = config
        self.provider = provider
        self.tools = ToolRegistry()
        self.extensions = ExtensionManager()
        self.session = Session(
            storage=InMemoryStorage(),
            max_context_tokens=config.max_context_tokens,
        )
        self._running = False
        self._abort_controller = asyncio.Event()

    # ------------------------------------------------------------------
    # Registration
    # ------------------------------------------------------------------

    def register_tool(self, tool: BaseTool) -> None:
        """Register a tool."""
        self.tools.register(tool)

    def register_extension(self, ext: Extension) -> None:
        """Register an extension."""
        self.extensions.register(ext)

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def start(self) -> None:
        """Start the agent."""
        self._running = True
        self._abort_controller.clear()

        # Signal handlers
        for sig in (signal.SIGINT, signal.SIGTERM):
            signal.signal(sig, self._handle_signal)

        self.extensions.execute_hook(HookPoint.ON_START)
        print(f"[{self.config.name}] Agent started")

    async def stop(self) -> None:
        """Stop the agent gracefully."""
        self._running = False
        self._abort_controller.set()
        self.extensions.execute_hook(HookPoint.ON_STOP)
        print(f"[{self.config.name}] Agent stopped")

    def _handle_signal(self, signum: int, frame: Any) -> None:
        """Handle shutdown signals."""
        print(f"\n[{self.config.name}] Received signal {signum}, shutting down...")
        asyncio.get_event_loop().create_task(self.stop())

    # ------------------------------------------------------------------
    # Core Execution
    # ------------------------------------------------------------------

    async def run(self, user_input: str) -> str:
        """
        Process a single user input through the full agent loop.

        Flow:
        1. Add user message to session
        2. Build context from session history
        3. Run agent loop (LLM → tools → LLM → ...)
        4. Return final assistant response
        """
        if not self._running:
            await self.start()

        # 1. Record user message
        self.session.append_message("user", user_input)

        # 2. Build context
        messages = self.session.build_context(self.config.system_prompt)

        # 3. Pre-LLM hook
        self.extensions.execute_hook(HookPoint.BEFORE_LLM, {"messages": messages})

        # 4. Build agent context
        context = AgentContext(
            system_prompt=self.config.system_prompt,
            messages=[Message(role=m["role"], content=[TextContent(text=m["content"])]) for m in messages],
            tools=[],  # Tools are handled via the registry
        )

        # 5. Run agent loop
        new_messages = agent_loop(context)

        # 6. Record assistant messages
        final_response = ""
        for msg in new_messages:
            if msg.role == "assistant":
                for part in msg.content:
                    if isinstance(part, TextContent):
                        self.session.append_message("assistant", part.text)
                        final_response = part.text

        # 7. Post-LLM hook
        self.extensions.execute_hook(HookPoint.AFTER_LLM, {"response": final_response})

        return final_response

    async def run_interactive(self) -> None:
        """Run in interactive mode (stdin/stdout)."""
        await self.start()

        print(f"\n{self.config.name} — Interactive Mode")
        print("Type 'quit' to exit, 'history' to see conversation\n")

        while self._running:
            try:
                user_input = input("You: ").strip()
            except (EOFError, KeyboardInterrupt):
                break

            if not user_input:
                continue

            if user_input.lower() == "quit":
                break

            if user_input.lower() == "history":
                history = self.session.get_history()
                for entry in history:
                    data = entry.data
                    print(f"  [{entry.type.value}] {data.get('content', data)}")
                continue

            response = await self.run(user_input)
            print(f"Agent: {response}\n")

        await self.stop()

    # ------------------------------------------------------------------
    # State
    # ------------------------------------------------------------------

    def get_state(self) -> dict[str, Any]:
        """Get current agent state."""
        return {
            "config": {
                "name": self.config.name,
                "system_prompt": self.config.system_prompt[:100] + "...",
            },
            "tools": self.tools.list_tools(),
            "extensions": self.extensions.list_extensions(),
            "session": self.session.get_checkpoint(),
            "running": self._running,
        }


# ============================================================================
# Example Usage
# ============================================================================

async def main():
    # Create agent
    config = AgentConfig(
        name="my-agent",
        system_prompt="You are a helpful coding assistant. Use tools to read and write files.",
    )

    provider_config = ProviderConfig(
        name="openai",
        api_key="sk-...",
        model="gpt-4o",
    )

    harness = AgentHarness(config=config, provider=OpenAIProvider(provider_config))

    # Register tools
    from minimal_tool import ReadFileTool, WriteFileTool
    harness.register_tool(ReadFileTool())
    harness.register_tool(WriteFileTool())

    # Register extensions
    from minimal_extension import LoggingExtension, TimingExtension
    harness.register_extension(LoggingExtension())
    harness.register_extension(TimingExtension())

    # Print state
    import json
    print("=== Agent State ===")
    print(json.dumps(harness.get_state(), indent=2, default=str))

    # Run interactive
    await harness.run_interactive()


if __name__ == "__main__":
    asyncio.run(main())
